### PR TITLE
[Rbac] Alcor Rbac Support

### PR DESCRIPTION
**Pull request by Gzure**
This PR support rbac policy.

### CODE REVIEW

File: web/src/main/java/com/futurewei/alcor/web/rbac/aspect/Rbac.java [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+ */
+
+package com.futurewei.alcor.web.rbac.aspect;
+
+import java.lang.annotation.*;
+
+@Target(ElementType.METHOD)
+@Retention(RetentionPolicy.RUNTIME)
+@Documented
+public @interface Rbac {
```
[the remark this answers is not in the document]
Comment (Gzure) [reply web/src/main/java/com/futurewei/alcor/web/rbac/aspect/Rbac.java:26]: Yes, It's an annotation.

**Comment by xieus** [comment]
@Gzure Can you draft a short design spec to explain the high-level design points on Alcor RBAC? Thanks.

**Comment by Gzure** [comment]
> @Gzure Can you draft a short design spec to explain the high-level design points on Alcor RBAC? Thanks.

@xieus I add a simple rbac design file [https://github.com/Gzure/alcor/blob/subnet-delete-bug/docs/modules/ROOT/pages/high_level/rbac_design.adoc](rbac_design.adoc)

File: lib/src/main/java/com/futurewei/alcor/common/rbac/RbacManagerConfiguration.java [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+
+import java.io.IOException;
+
+@Configuration
+public class RbacManagerConfiguration {
+
+    @Value("${rbac.policy.type: strict}")
+    private String rbacType;
```
Comment (xieus) [inline lib/src/main/java/com/futurewei/alcor/common/rbac/RbacManagerConfiguration.java:31]: What other rbacType do we plan to support?

File: lib/src/main/java/com/futurewei/alcor/common/rbac/StrictRbacManager.java [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+import org.springframework.context.annotation.Configuration;
+import org.springframework.stereotype.Component;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.lang.reflect.Field;
+import java.util.*;
+
+public class StrictRbacManager implements RbacManger {
+    private static final Logger LOG = LoggerFactory.getLogger(StrictRbacManager.class);
```
Comment (xieus) [inline lib/src/main/java/com/futurewei/alcor/common/rbac/StrictRbacManager.java:39]: We have also a Alcor logger under common lib. do you find anything missing compared to slf4j? If not, maybe a good idea to switch to our own logger :-)

File: lib/src/main/java/com/futurewei/alcor/common/rbac/StrictRbacManager.java [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+    private static final String ADMIN_ROLE_NAME = "admin";
+    private static final String ADVSVC_ROLE_NAME = "advsvc";
+
+    private ServiceRbacRule serviceRbacRule;
+
+    public StrictRbacManager() throws IOException {
+        initServiceRbacRule();
+    }
+
+    private void initServiceRbacRule() throws IOException {
```
Comment (xieus) [inline lib/src/main/java/com/futurewei/alcor/common/rbac/StrictRbacManager.java:53]: initServiceRbacRules

File: web/src/main/java/com/futurewei/alcor/web/rbac/aspect/RbacConfiguration.java [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+
+package com.futurewei.alcor.web.rbac.aspect;
+
+import org.springframework.context.annotation.ComponentScan;
+import org.springframework.context.annotation.Configuration;
+
+@Configuration
+@ComponentScan(value = {"com.futurewei.alcor.common.rbac", "com.futurewei.alcor.web.rbac.aspect"})
+public class RbacConfiguration {
+}
```
Comment (xieus) [inline web/src/main/java/com/futurewei/alcor/web/rbac/aspect/RbacConfiguration.java:27]: Do we plan to have some fun with RbacConfiguration? It is empty for now. :-)

File: web/src/main/java/com/futurewei/alcor/web/rbac/aspect/Rbac.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (xieus) [reply web/src/main/java/com/futurewei/alcor/web/rbac/aspect/Rbac.java:26]: Can you document this new annotation in the design doc? This is good staff.

File: web/src/main/java/com/futurewei/alcor/web/rbac/aspect/RbacAspect.java [hunk 6]
```diff
@@ -0,0 +1,238 @@
...
+import java.lang.annotation.Annotation;
+import java.lang.reflect.Field;
+import java.lang.reflect.Method;
+import java.util.*;
+
+import static com.futurewei.alcor.common.constants.CommonConstants.*;
+
+@Aspect
+@Component
+public class RbacAspect {
```
Comment (xieus) [inline web/src/main/java/com/futurewei/alcor/web/rbac/aspect/RbacAspect.java:53]: Pretty important class. Some more comments would be helpful.

File: web/src/main/java/com/futurewei/alcor/web/rbac/aspect/RbacAspect.java [hunk 7]
```diff
@@ -0,0 +1,238 @@
...
+
+    @Pointcut("@annotation(com.futurewei.alcor.web.rbac.aspect.Rbac)")
+    public void annotationPointCut(){}
+
+    @Around("annotationPointCut()")
+    public Object checkRbac(ProceedingJoinPoint pjp) throws Throwable {
+
+        String methodType = request.getMethod().toUpperCase();
+        String tokenInfo = request.getHeader(TOKEN_INFO_HEADER);
+        MethodSignature ms = (MethodSignature)pjp.getSignature();
```
Comment (xieus) [inline web/src/main/java/com/futurewei/alcor/web/rbac/aspect/RbacAspect.java:78]: Could ms be possibly NULL or guaranteed to have value?

File: docs/modules/ROOT/pages/high_level/rbac_design.adoc [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+
+== WorkFlow
+Rbac check around a http request.
+
+. a request come
+. rbac check method access, failed throw exception
+. rbac check field access, failed throw exception
+. handle request
+. rbac check field access, failed hide the field
+. response the request
```
Comment (xieus) [inline docs/modules/ROOT/pages/high_level/rbac_design.adoc:33]: In this design doc, I would recommend to add a few more sections
- Microservices with support of Rbac (which miroservices need RBAC and why...)
- Authorization implementation (is the validation local on a microservice or need to talk to KeyStone etc.)
- An example rbac_policy.json and a brief explanation.

File: docs/modules/ROOT/pages/high_level/high_level_design.adoc [hunk 9]
```diff
@@ -1,63 +1 @@
...
=== Compatibility

* Compatibility with OpenStack, Kubernetes, and OPNFV
* Support multiple data plane implementations including Mizar (eBPF/Geneve) currently and OVS/VxLan (in design)
** Open driver framework at Controllers
** Plugin model at Alcor Control Agent

== Inside the Big Picture

plantuml::../../workflows/vm_create.puml[format="svg", align="center"]
```
[the remark this answers is not in the document]
Comment (Gzure) [reply docs/modules/ROOT/pages/high_level/high_level_design.adoc:1]: > This file appears to be changed unintentionally. Could you revert it back?

ok

File: lib/src/main/java/com/futurewei/alcor/common/rbac/RbacManagerConfiguration.java (hunk 2 above)
Comment (Gzure) [reply lib/src/main/java/com/futurewei/alcor/common/rbac/RbacManagerConfiguration.java:31]: There are two rbacType Ignore or strict. Ignore rbac let all requestes access.

File: lib/src/main/java/com/futurewei/alcor/common/rbac/StrictRbacManager.java (hunk 3 above)
Comment (Gzure) [reply lib/src/main/java/com/futurewei/alcor/common/rbac/StrictRbacManager.java:39]: ok

File: web/src/main/java/com/futurewei/alcor/web/rbac/aspect/RbacConfiguration.java (hunk 5 above)
Comment (Gzure) [reply web/src/main/java/com/futurewei/alcor/web/rbac/aspect/RbacConfiguration.java:27]: This file is config @ComponentScan Only, So it have empty body.

File: docs/modules/ROOT/pages/high_level/rbac_design.adoc [hunk 10]
```diff
@@ -0,0 +1,33 @@
+= Alcor Rbac Design
+v0.1, 2020-08-14
+:toc: right
+:imagesdir: ../../images
+
+== Introduction
+Alcor Rbac (Role based access control) is similar to neutron Rbac. It's base on Openstack Keystone roles.
```
[the remark this answers is not in the document]
Comment (Gzure) [reply docs/modules/ROOT/pages/high_level/rbac_design.adoc:7]: > Could you give a reference link here for "OpenStack Keystone roles"

[Openstack Keystone roles](https://docs.openstack.org/api-ref/identity/v3/?expanded=list-roles-detail#roles)

File: web/src/main/java/com/futurewei/alcor/web/rbac/aspect/RbacAspect.java (hunk 7 above)
Comment (Gzure) [reply web/src/main/java/com/futurewei/alcor/web/rbac/aspect/RbacAspect.java:78]: > Could ms be possibly NULL or guaranteed to have value?
Ms could not be null.

File: lib/src/main/java/com/futurewei/alcor/common/rbac/StrictRbacManager.java (hunk 3 above)
Comment (Gzure) [reply lib/src/main/java/com/futurewei/alcor/common/rbac/StrictRbacManager.java:39]: > We have also a Alcor logger under common lib. do you find anything missing compared to slf4j? If not, maybe a good idea to switch to our own logger :-)

Slf4j is a stateful and strong log frame. All our own logger features can supported by Slf4j.

File: docs/modules/ROOT/pages/high_level/rbac_design.adoc [hunk 11]
```diff
@@ -0,0 +1,227 @@
...
+          "rule_type": "admin_only",
+          "roles": ["admin"]
+        }
+      ]
+    },
+    {
+      "name": "get_vpc",
+      "method_type": "GET",
+      "description": "Get or list vpc",
+      "rule_type": "admin_or_owner",
```
Comment (xieus) [inline docs/modules/ROOT/pages/high_level/rbac_design.adoc:126]: @Gzure if rule_type=admin_or_owner, why is the value of roles required to be admin? What would be the roles of the token included in our X-Token-info header field?

File: lib/src/main/java/com/futurewei/alcor/common/rbac/RbacManagerConfiguration.java (hunk 2 above)
Comment (xieus) [reply lib/src/main/java/com/futurewei/alcor/common/rbac/RbacManagerConfiguration.java:31]: Okay thanks. In this case, we would suggest to name the rbacType as Ignore and Enforced.

File: docs/modules/ROOT/pages/high_level/rbac_design.adoc (hunk 10 above)
[the remark this answers is not in the document]
Comment (xieus) [reply docs/modules/ROOT/pages/high_level/rbac_design.adoc]: I mean adding the reference to the design doc :-)

File: docs/modules/ROOT/pages/high_level/rbac_design.adoc (hunk 11 above)
Comment (Gzure) [reply docs/modules/ROOT/pages/high_level/rbac_design.adoc:126]: > @Gzure if rule_type=admin_or_owner, why is the value of roles required to be admin? What would be the roles of the token included in our X-Token-info header field?
Admin_or_owner rule type will ignore the roles.

Comment (xieus) [reply docs/modules/ROOT/pages/high_level/rbac_design.adoc:126]: Okay in this case, please keep the example as accurate as possible. maybe remove any entry in the roles array. @Gzure

Comment (Gzure) [reply docs/modules/ROOT/pages/high_level/rbac_design.adoc:126]: > Okay in this case, please keep the example as accurate as possible. maybe remove any entry in the roles array. @Gzure

ok

Comment (Gzure) [reply docs/modules/ROOT/pages/high_level/rbac_design.adoc:126]: > Okay in this case, please keep the example as accurate as possible. maybe remove any entry in the roles array. @Gzure

Okay, I will remove the roles.

**Pull request merged**